File: semaphore.cpp
#include <iostream>
#include <thread>
#include <atomic>
using namespace std;

void producer();
void consumer();

void wait();
void signal();

int total = 10;
constexpr int count = 100000;

atomic<int> S(1);

int main(void) {
    cout << "Start : " << total << endl;

    thread t1(producer);
    thread t2(consumer);

    t1.join();
    t2.join();

    cout << "End : " << total << endl;

    return 0;
}

void producer() {
    for (int i = 0; i < count; i++) {
        wait();
        total++;
        signal();
    }
}

void consumer() {
    for (int i = 0; i < count; i++) {
        wait();
        total--;
        signal();
    }
}

void wait() {
    while (true) {
        int s = S.load();
        if (s > 0 && S.compare_exchange_strong(s, s - 1))
            break;
    }
}

void signal() {
    S.fetch_add(1);
}

File: Semaphore.java
import java.util.concurrent.atomic.AtomicInteger;

public class Semaphore {
    public static int total = 10;
    public static final int LOOPS = 100000;

    public static final AtomicInteger S = new AtomicInteger(1);

    public static void main(String[] args) throws InterruptedException {
        System.out.println("Start : " + total);

        Thread producer = new Thread(new Producer());
        Thread consumer = new Thread(new Consumer());

        producer.start();
        consumer.start();

        producer.join();
        consumer.join();

        System.out.println("End : " + total);
    }

    public static void _wait() {
        while (true) {
            int s = S.get();

            if (s > 0 && S.compareAndSet(s, s - 1))
                break;
        }
    }

    public static void _signal() {
        S.incrementAndGet();
    }

    public static class Producer implements Runnable {
        @Override
        public void run() {
            for (int i = 0; i < LOOPS; i++) {
                _wait();
                total++;
                _signal();
            }
        }
    }

    public static class Consumer implements Runnable {
        @Override
        public void run() {
            for (int i = 0; i < LOOPS; i++) {
                _wait();
                total--;
                _signal();
            }
        }
    }
}
